Bump citus to 10.2.0 (#5276)
citusdata/citus: CHANGELOG.md

### citus v10.2.0 (September 11, 2021) ###

* Fixes a bug that may cause crash while aborting transaction

* Fixes bug about int and smallint sequences on MX

> * Fixes a bug that caused worker_append_table_to_shard to write as superuser

Reviewer: ```suggestion
* Fixes a bug that caused `worker_append_table_to_shard` to write as superuser
```

> * Enable transferring of data using binary encoding by default on PG14

Reviewer: ```suggestion
* Enables transferring of data using binary encoding by default on PG14
```

> * Add helper UDFs for easy time partition management

Reviewer: should we mention UDF names here, I think so

Reviewer 2: Added those in a multiline entry

> * Adds pg14 support

Reviewer: obviously, should be the top element, just noting

> * Columnar: enable chunk group filtering to work with Params.

Reviewer: ```suggestion
* Enables chunk group filtering to work with Params for columnar tables
```

Reviewer: Should also have, in addition to above^:

"Enables qual push down for joins including columnar tables"

TODO: PLEASE SHORTEN THE NEXT LINE MANUALLY, IT SHOULD BE NO LONGER THAN 78 CHARS
* Fixes a bug that prevents attaching partitions when colocated foreign key exists

* Speeds up dropping partitioned tables

> * Fix crash in shard rebalancer when no distributed tables exist

Reviewer: ```suggestion
* Fixes a crash in shard rebalancer when no distributed tables exist
```

> * Fix a segfault caused by use after free in ConnectionsPlacementHash

Reviewer: ```suggestion
* Fixes a segfault caused by use after free in when using a cached connection
```

TODO: PLEASE SHORTEN THE NEXT LINE MANUALLY, IT SHOULD BE NO LONGER THAN 78 CHARS

> * Acquire AccessShareLock before updating table statistics to verify the table exists

Reviewer: ```suggestion
* Makes sure that table exists before updating table statistics
```

* Allows more graceful failovers when replication factor > 1

* Fixes a bug that cause partitions to have wrong dist. key after DROP COLUMN

> * Columnar: fix misleading comments and useless types.

Reviewer: ```suggestion
```

> * Fix master_update_table_statistics scripts for 9.4, 9.5, 10.0, 10.1

Reviewer: should this be "Improves master_update_table_statistics and provides distributed deadlock detection" as in the older versions, or was it something different ?

Reviewer 2: Yes, that is correct.

We updated the description on the changelog, but the scripts fetched the former version of these entries.

TODO: PLEASE SHORTEN THE NEXT LINE MANUALLY, IT SHOULD BE NO LONGER THAN 78 CHARS

> * Reduces memory usage of columnar table scans by freeing the memory used for last stripe rea

Reviewer: ```suggestion
* Reduces memory usage of columnar table scans by freeing the memory used for last stripe read
```

> * Fix showing target shard size in the rebalance progress monitor

Reviewer: ```suggestion
* Fixes showing target shard size in the rebalance progress monitor
```

> * Include data_type and cache in sequence definition on worker

Reviewer: ```suggestion
* Includes `data_type` and `cache` in sequence definition on worker
```

> ### citus v10.2.0 (September 11, 2021) ###

Reviewer 3: Is the date OK (3 days ago)?

Reviewer 2: I updated it to today's date

* Adds PostgreSQL 14 support

* Adds hash & btree index support for columnar tables

* Adds helper UDFs for easy time partition management:
  `get_missing_time_partition_ranges`, `create_time_partitions`, and
  `drop_old_time_partitions`

* Adds propagation of ALTER SEQUENCE

* Adds support for ALTER INDEX ATTACH PARTITION

* Adds support for CREATE INDEX ON ONLY

* Allows more graceful failovers when replication factor > 1

* Enables chunk group filtering to work with Params for columnar tables

* Enables qual push down for joins including columnar tables

* Enables transferring of data using binary encoding by default on PG14

* Improves `master_update_table_statistics` and provides distributed deadlock
  detection

* Includes `data_type` and `cache` in sequence definition on worker

> * Makes sure that table exists before updating table statistics

Reviewer 3: I think the only PR missing is https://github.com/citusdata/citus/pull/5144

I just added description there, can you move here as well?

  last stripe read

* Shows projected columns for columnar tables in EXPLAIN output

* Speeds up dropping partitioned tables

* Synchronizes hasmetadata flag on mx workers

* Uses current user while syncing metadata

* Fixes a bug about int and smallint sequences on MX

* Fixes a bug that cause partitions to have wrong distribution key after
  DROP COLUMN

* Fixes a bug that caused `worker_append_table_to_shard` to write as superuser

* Fixes a bug that caused `worker_create_or_alter_role` to crash with NULL input

* Fixes a bug that causes pruning incorrect shard of a range distributed table

* Fixes a bug that may cause crash while aborting transaction

* Fixes a bug that prevents attaching partitions when colocated foreign key
  exists

* Fixes a bug with `nextval('seq_name'::text)`

* Fixes a crash in shard rebalancer when no distributed tables exist

* Fixes a segfault caused by use after free in when using a cached connection

* Fixes a UNION pushdown issue

* Fixes a use after free issue that could happen when altering a distributed
  table

* Fixes showing target shard size in the rebalance progress monitor

> * Adds a parameter to cleanup metadata when metadata syncing is stopped

Reviewer 3: minor: probably should go right before `Fixes ..` items

### citus v10.2.0 (September 11, 2021) ###

* Adds PostgreSQL 14 support

* Adds hash & btree index support for columnar tables

* Adds helper UDFs for easy time partition management:
  `get_missing_time_partition_ranges`,
  `create_time_partitions`,
  `drop_old_time_partitions`

> * Adds a parameter to cleanup metadata when metadata syncing is stopped

Reviewer 3: minor: I'd move this as the last item with `Adds ..`

Reviewer 2: moved at the end, after the `Fixes ..` items

* Adds propagation of ALTER SEQUENCE

* Adds support for ALTER INDEX ATTACH PARTITION

* Adds support for CREATE INDEX ON ONLY

* Allows more graceful failovers when replication factor > 1

* Enables chunk group filtering to work with Params for columnar tables

* Enables qual push down for joins including columnar tables

* Enables transferring of data using binary encoding by default on PG14

* Fixes a bug that allowed moving of shards belonging to a reference table

* Fixes a bug about int and smallint sequences on MX

* Fixes a bug that cause partitions to have wrong distribution key after
  DROP COLUMN

> * Fixes a bug that caused `worker_append_table_to_shard` to write as superuser

Reviewer 3: We also backported this: https://github.com/citusdata/citus/pull/5256  but never released earlier versions.

What is our policy on such cases. We just keep the changelog entry here?

Reviewer 2: I do not have enough context on this PR.

My personal opinion on what to include on `10.2.0` is the following: Add all entries that are introduced after `10.1.0` including those that are already listed under `10.1.1` and `10.1.2`. This way the user can only take a look at the major versions to understand about the capability of each version.

For example, the end user may not understand that all the fixes in `10.1.1` are also shipped in `10.2.0`. We as Citus developers know about this pattern, but I am not so sure if the end users feel the same way.

Reviewer: If we didn't release yet, then I think we should have changelog entry here.
Otherwise, we don't put it to the new release's changelog entries.

Reviewer 2: If a bug never existed on any earlier version, then it makes more sense to drop the entry for the bugfix. Though I am not so sure if that's the case for this PR.

It would be nice if we can have `cherry-pick-X.Y` on the PR to understand which versions are actually affected.

Reviewer 3: This is back ported up to 9.4 (including enterprise): https://github.com/citusdata/citus/commits/release-9.4

I guess our typical approach is to have this entry on 9.4.X change logs? I think to be consistent, I have a slight preference to have it only on 9.4.X

Reviewer 2: > I guess our typical approach is to have this entry on 9.4.X change logs?

We usually add the item to all the versions that we backported the fix to.

For example, a user that uses 10.0 will be able to see if a patch release can solve an issue that they run into.

For example the following entry is in `9.4.5`, `9.5.5`, `10.0.4` and `10.1.0` and it makes sense to me.
> Fixes an issue that could cause citus_finish_pg_upgrade to fail

Reviewer 3: Alright, then lets keep it here. Thanks for the explanation

### citus v10.2.0 (September 11, 2021) ###

* Adds pg14 support

* Acquires AccessShareLock before updating table statistics to verify the table
  exists

* Adds a parameter to cleanup metadata

* Adds basic index support for columnar tables

* Adds hash & btree index support for columnar tables

* Adds helper UDFs for easy time partition management:
  `get_missing_time_partition_ranges`,
  `create_time_partitions`,

>   `drop_old_time_partitions`

Reviewer: let's make this one of top items, after pg14 support

* Adds propagation of ALTER SEQUENCE

* Adds support for ALTER INDEX ATTACH PARTITION

* Adds support for CREATE INDEX ON ONLY

* Allows more graceful failovers when replication factor > 1

* Enables chunk group filtering to work with Params for columnar tables

* Enables qual push down for joins including columnar tables

* Enables transferring of data using binary encoding by default on PG14

* Fixes a bug that allowed moving of shards belonging to a reference table

* Fixes a bug about int and smallint sequences on MX

* Fixes a bug that cause partitions to have wrong distribution key after
  DROP COLUMN

* Fixes a bug that caused `worker_append_table_to_shard` to write as superuser

* Fixes a bug that caused `worker_create_or_alter_role` to crash with NULL input

* Fixes a bug that causes pruning incorrect shard of a range distributed table

* Fixes a bug that may cause crash while aborting transaction

* Fixes a bug that prevents attaching partitions when colocated foreign key
  exists

* Fixes a bug where foreign key to reference table was disallowed

> * Fixes a bug with local cached plans on tables with dropped columns

Reviewer 2: No we do not need this one. I removed it

* Fixes a bug with `nextval('seq_name'::text)`

* Fixes a crash in shard rebalancer when no distributed tables exist

* Fixes a segfault caused by use after free in when using a cached connection

* Fixes a UNION pushdown issue

* Fixes a use after free issue that could happen when altering a distributed
  table

> * Fixes an issue that could cause `citus_finish_pg_upgrade` to fail

Reviewer 2: Removed.

* Fixes bug about int and smallint sequences on MX

> * Fixes showing target shard size in the rebalance progress monitor

Reviewer 3: This is already in Citus 10.1? https://github.com/citusdata/citus/pull/5136

Reviewer 2: Yes it is in 10.1.1. Removed.

* Fixes the relation size bug during rebalancing

* Fixes two race conditions in the `get_rebalance_progress`

> * Improves `master_update_table_statistics` and provides distributed deadlock

Reviewer 3: I think we can move these items right before `Fixes..` change log entries. These are features that feels better suited before the `fixes` items?

Reviewer 2: I moved the fixes to the very bottom of the list. Do you think it looks good now?

  detection

* Includes `data_type` and `cache` in sequence definition on worker

* Makes sure that table exists before updating table statistics

* Reduces memory usage of columnar table scans by freeing the memory used for
  last stripe read

* Shows projected columns for columnar tables in EXPLAIN output

* Speeds up dropping partitioned tables

* Synchronizes hasmetadata flag on mx workers

* Uses current user while syncing metadata

> 

Reviewer: for example, most of those should be placed before the bugfixes

Reviewer 2: I placed `Fixes ..` items at the very bottom. Does that look good now?

* Adds pg14 support

* Acquires AccessShareLock before updating table statistics to verify the table exists

* Adds a parameter to cleanup metadata

* Adds basic index support for columnar tables

> * Adds hash & btree index support for columnar tables

Reviewer: duplicate

Reviewer: should remove the first one

Reviewer: let's make this one of top items, after pg14 support
